firmware/repository.py

```diff
@@ -0,0 +1,126 @@
+"""Repository - module to communicate with the database"""
+
+from firmware.database import db_session
+from firmware.models import User, Company, Category, Review
+from sqlalchemy import and_
+
+
+def get_companies():
+    """ gets all companies from database """
+    companies = db_session.query(Company)
+    return companies
+
+
+def get_user_id(username):
+    """ gets a user id via it's unique username """
+    user_id = db_session.query(User.id).filter(User.username == username)
+    return user_id
+
+
+def add_company(company):
+    """ adds a company to the database """
+    db_session.add(company)
+    db_session.commit()
+    return company.id
+
+
+def get_company(company_id):
+    """ retrieves a specific company, via it's id """
+    company = db_session.query(Company).filter(Company.id == company_id).one()
+    if company is None:
+        return None
+    return company
+
+
+def get_category(company_id):
+    """ gets a category for a specific company """
+    company = db_session.query(Company).filter(Company.id == company_id).one()
+    category = db_session.query(Category).filter(company.category_id ==
+                                                 Category.id).one()
+    return category.domain
+
+
+def get_categories():
+    """ gets all categories from the database """
+    categories = db_session.query(Category).all()
+    return categories
+
+
+def get_reviews(company_id):
+    """ gets all reviews for a company """
+    reviews = db_session.query(Review).filter(Review.company_id == company_id)\
+                .order_by(Review.id.desc()).all()
+    return reviews
+
+
+def add_user(user):
+    """ adds a new user to the database """
+    db_session.add(user)
+    db_session.commit()
+    return user.id, user.username
+
+
+def get_user(username):
+    """gets a user by unique username """
+    user = db_session.query(User).filter(User.username == username).one().serialize()
```

Comment:
What does `serialize()` do here?

Reply:
Resolves "TypeError: <firmware.models.User object at 0x1019f2710> is not JSON serializable"

Trace:
File "/Library/Python/2.7/site-packages/flask/app.py", line 1997, in __call__
return self.wsgi_app(environ, start_response)
File "/Library/Python/2.7/site-packages/flask/app.py", line 1985, in wsgi_app
response = self.handle_exception(e)
File "/Library/Python/2.7/site-packages/flask/app.py", line 1540, in handle_exception
reraise(exc_type, exc_value, tb)
File "/Library/Python/2.7/site-packages/flask/app.py", line 1982, in wsgi_app
response = self.full_dispatch_request()
File "/Library/Python/2.7/site-packages/flask/app.py", line 1615, in full_dispatch_request
return self.finalize_request(rv)
File "/Library/Python/2.7/site-packages/flask/app.py", line 1632, in finalize_request
response = self.process_response(response)
File "/Library/Python/2.7/site-packages/flask/app.py", line 1858, in process_response
self.save_session(ctx.session, response)
File "/Library/Python/2.7/site-packages/flask/app.py", line 924, in save_session
return self.session_interface.save_session(self, session, response)
File "/Library/Python/2.7/site-packages/flask/sessions.py", line 363, in save_session
val = self.get_signing_serializer(app).dumps(dict(session))
File "/Library/Python/2.7/site-packages/itsdangerous.py", line 565, in dumps
payload = want_bytes(self.dump_payload(obj))
File "/Library/Python/2.7/site-packages/itsdangerous.py", line 847, in dump_payload
json = super(URLSafeSerializerMixin, self).dump_payload(obj)
File "/Library/Python/2.7/site-packages/itsdangerous.py", line 550, in dump_payload
return want_bytes(self.serializer.dumps(obj))
File "/Library/Python/2.7/site-packages/flask/sessions.py", line 85, in dumps
return json.dumps(_tag(value), separators=(',', ':'))
File "/Library/Python/2.7/site-packages/flask/json.py", line 123, in dumps
rv = _json.dumps(obj, **kwargs)
File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 250, in dumps
sort_keys=sort_keys, **kw).encode(obj)
File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 209, in encode
chunks = list(chunks)
File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 434, in _iterencode
for chunk in _iterencode_dict(o, _current_indent_level):
File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
for chunk in chunks:
File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 442, in _iterencode
o = _default(o)
File "/Library/Python/2.7/site-packages/flask/json.py", line 80, in default
return _json.JSONEncoder.default(self, o)
File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 184, in default
Open an interactive python shell in this frameraise TypeError(repr(o) + " is not JSON serializable")